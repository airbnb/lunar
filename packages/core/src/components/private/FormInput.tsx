import React from 'react';
import { Omit } from 'utility-types';
import withStyles, { css, WithStylesProps } from '../../composers/withStyles';
import buildInputStyles from '../../themes/buildInputStyles';

export type IgnoreAttributes =
  | 'children'
  | 'dangerouslySetInnerHTML'
  | 'onChange'
  | 'value'
  // RDFa attributes
  | 'about'
  | 'datatype'
  | 'inlist'
  | 'prefix'
  | 'property'
  | 'resource'
  | 'typeof'
  | 'vocab'
  // Non-standard attributes
  | 'inputMode'
  | 'is'
  | 'radioGroup'
  | 'color'
  | 'results'
  | 'security';

export type Props<T = any> = {
  /** Decrease font size and padding. */
  compact?: boolean;
  /** Mark the field as important. */
  important?: boolean;
  /** Mark the field as invalid. */
  invalid?: boolean;
  /** Add "notranslate" className to prevent Google Chrome translation. */
  noTranslate?: boolean;
  /** Mark the field as optional. */
  optional?: boolean;
  /** Current value. */
  value?: string;
  /** Callback to access the underlying input DOM element. */
  wrappedRef?: React.Ref<T>;
};

export type InputProps = Omit<React.InputHTMLAttributes<HTMLInputElement>, IgnoreAttributes> &
  Props<HTMLInputElement>;

export type SelectProps = Omit<React.SelectHTMLAttributes<HTMLSelectElement>, IgnoreAttributes> &
  Props<HTMLSelectElement>;

export type TextAreaProps = Omit<
  React.TextareaHTMLAttributes<HTMLTextAreaElement>,
  IgnoreAttributes
> &
  Props<HTMLTextAreaElement>;

export type PrivateProps = Props &
  WithStylesProps & {
    // Support everything for convenience
    [key: string]: any;
    children?: React.ReactNode;
    hasPrefix?: boolean;
    hasSuffix?: boolean;
    tagName: 'input' | 'select' | 'textarea';
  };

class FormInput extends React.Component<PrivateProps> {
  static defaultProps = {
    children: null,
    compact: false,
    disabled: false,
    hasPrefix: false,
    hasSuffix: false,
    hidden: false,
    important: false,
    invalid: false,
    noTranslate: false,
    optional: false,
    value: '',
  };

  render() {
    const {
      children,
      compact,
      disabled,
      hasPrefix,
      hasSuffix,
      hidden,
      id,
      important,
      invalid,
      noTranslate,
      wrappedRef,
      optional,
      styles,
      tagName: Tag,
      ...restProps
    } = this.props;
    const isSelect = Tag === 'select';
    const props: any = {
      ...restProps,
      ...css(
        styles.input,
        important && styles.input_important,
        compact && styles.input_compact,
        invalid && styles.input_invalid,
        disabled && styles.input_disabled,
        hidden && styles.input_hidden,
        isSelect && styles.select,
        isSelect && compact && styles.select_compact,
        hasPrefix && styles.input_hasPrefix,
        hasSuffix && styles.input_hasSuffix,
      ),
      disabled,
      id,
      required: !optional,
    };

    // Only populate when invalid, otherwise it will break some CSS selectors
    if (invalid) {
      props['aria-invalid'] = true;
      props['aria-describedby'] = `${id}-error`;
    }

    // Cannot use chidren for input/textarea as they are void elements
    if (isSelect) {
      props.children = children;
    }

    // Add magical className to prevent Google Chrome translation
    if (noTranslate) {
      props.className += ' notranslate';
    }

    // @ts-ignore [ts] JSX element type 'Component' does not have any construct or call signatures. [2604]
    return <Tag {...props} ref={wrappedRef} data-gramm="false" data-enable-grammarly="false" />;
  }
}

export default withStyles(buildInputStyles)(FormInput);
