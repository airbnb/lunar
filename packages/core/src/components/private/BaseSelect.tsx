import React from 'react';
import IconCaretDown from '@airbnb/lunar-icons/lib/interface/IconCaretDown';
import withStyles, { WithStylesProps } from '../../composers/withStyles';
import FormInput, { SelectProps } from './FormInput';

export type Props = SelectProps & {
  /** List of `option`s to render. */
  children: NonNullable<React.ReactNode>;
  /** An empty `option` to render at the top of the list. */
  placeholder?: string;
  /** Callback fired when the value changes. */
  onChange: (value: string, event: React.ChangeEvent<HTMLSelectElement>) => void;
};

class BaseSelect extends React.Component<Props & WithStylesProps> {
  static defaultProps = {
    placeholder: '',
  };

  private handleChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
    this.props.onChange(event.currentTarget.value, event);
  };

  render() {
    const { cx, children, placeholder, styles, ...restProps } = this.props;

    return (
      <div className={cx(styles.select)}>
        <FormInput {...restProps} onChange={this.handleChange} tagName="select">
          {placeholder && (
            <option value="" disabled>
              {placeholder}
            </option>
          )}

          {children}
        </FormInput>

        <div
          className={cx(
            styles.arrow,
            restProps.disabled && styles.arrow_disabled,
            restProps.invalid && styles.arrow_invalid,
            restProps.compact && styles.arrow_compact,
          )}
        >
          <IconCaretDown decorative size="2em" />
        </div>
      </div>
    );
  }
}

export default withStyles(({ pattern, unit }) => ({
  select: {
    position: 'relative',
    display: 'block',
    width: '100%',
  },

  arrow: {
    position: 'absolute',
    right: unit,
    top: '50%',
    transform: 'translateY(-50%)',
    pointerEvents: 'none',
    zIndex: 1,
  },

  arrow_disabled: {
    ...pattern.disabled,
  },

  arrow_invalid: {
    ...pattern.invalid,
  },

  arrow_compact: {
    right: unit * 0.75,
  },
}))(BaseSelect);
