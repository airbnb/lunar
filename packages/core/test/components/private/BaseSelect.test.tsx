import React from 'react';
import { shallowWithStyles } from '@airbnb/lunar-test-utils';
import BaseSelect from '../../../src/components/private/BaseSelect';
import FormInput from '../../../src/components/private/FormInput';

describe('<BaseSelect />', () => {
  it('renders an input with the correct field', () => {
    const wrapper = shallowWithStyles(
      <BaseSelect name="foo" onChange={() => {}}>
        <option value="bar">Bar</option>
      </BaseSelect>,
    );

    expect(wrapper.find(FormInput).prop('tagName')).toBe('select');
  });

  it('renders disabled', () => {
    const wrapper = shallowWithStyles(
      <BaseSelect name="foo" onChange={() => {}} disabled>
        <option value="bar">Bar</option>
      </BaseSelect>,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('renders invalid', () => {
    const wrapper = shallowWithStyles(
      <BaseSelect name="foo" onChange={() => {}} invalid>
        <option value="bar">Bar</option>
      </BaseSelect>,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('renders compact', () => {
    const wrapper = shallowWithStyles(
      <BaseSelect name="foo" onChange={() => {}} compact>
        <option value="bar">Bar</option>
      </BaseSelect>,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('renders options via children', () => {
    const option = <option value="bar">Bar</option>;
    const wrapper = shallowWithStyles(
      <BaseSelect name="foo" onChange={() => {}}>
        {option}
      </BaseSelect>,
    );

    expect(wrapper.contains(option)).toBe(true);
  });

  it('supports optgroup', () => {
    const wrapper = shallowWithStyles(
      <BaseSelect name="foo" onChange={() => {}}>
        <optgroup label="Group">
          <option value="bar">Bar</option>
        </optgroup>
        <option value="baz">Baz</option>
      </BaseSelect>,
    );

    expect(wrapper.find('optgroup')).toHaveLength(1);
    expect(wrapper.find('option')).toHaveLength(2);
  });

  it('supports a placeholder option', () => {
    const wrapper = shallowWithStyles(
      <BaseSelect name="foo" onChange={() => {}} placeholder="Select something">
        <option value="bar">Bar</option>
      </BaseSelect>,
    );

    expect(wrapper.find('option')).toHaveLength(2);
    expect(
      wrapper
        .find('option')
        .at(0)
        .equals(
          <option value="" disabled>
            Select something
          </option>,
        ),
    ).toBe(true);
  });

  it('triggers `onChange` handler', () => {
    const spy = jest.fn();
    const wrapper = shallowWithStyles(
      <BaseSelect name="foo" onChange={spy}>
        <option value="bar">Bar</option>
      </BaseSelect>,
    );
    const event = {
      currentTarget: {
        value: 'foo',
      },
    };

    wrapper.find(FormInput).simulate('change', event);

    expect(spy).toHaveBeenCalledWith('foo', event);
  });
});
