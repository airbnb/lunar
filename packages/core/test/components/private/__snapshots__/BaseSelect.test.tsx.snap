// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BaseSelect /> renders compact 1`] = `
<div
  className="select_jd2fw5"
>
  <FormInput
    compact={true}
    name="foo"
    onChange={[Function]}
    tagName="select"
  >
    <option
      value="bar"
    >
      Bar
    </option>
  </FormInput>
  <div
    className="arrow_1lvqbjq-o_O-arrow_compact_ib86un"
  >
    <IconCaretDown
      decorative={true}
      size="2em"
    />
  </div>
</div>
`;

exports[`<BaseSelect /> renders disabled 1`] = `
<div
  className="select_jd2fw5"
>
  <FormInput
    disabled={true}
    name="foo"
    onChange={[Function]}
    tagName="select"
  >
    <option
      value="bar"
    >
      Bar
    </option>
  </FormInput>
  <div
    className="arrow_1lvqbjq-o_O-arrow_disabled_5k6q5e"
  >
    <IconCaretDown
      decorative={true}
      size="2em"
    />
  </div>
</div>
`;

exports[`<BaseSelect /> renders invalid 1`] = `
<div
  className="select_jd2fw5"
>
  <FormInput
    invalid={true}
    name="foo"
    onChange={[Function]}
    tagName="select"
  >
    <option
      value="bar"
    >
      Bar
    </option>
  </FormInput>
  <div
    className="arrow_1lvqbjq-o_O-arrow_invalid_u10i2o"
  >
    <IconCaretDown
      decorative={true}
      size="2em"
    />
  </div>
</div>
`;
